pkg/gardenlet/controller/extensions/common_sync.go

```diff
@@ -0,0 +1,153 @@
+		extensionObject := newObj.(v1alpha1.Object)
+		if extensionObject.GetExtensionStatus().GetState() == nil {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueueOnUpdate(queue workqueue.RateLimitingInterface) func(newExtensionObject, oldExtensionObject interface{}) {
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(v1alpha1.Object)
+			oldExtensionObj = oldObj.(v1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
+		}
+
+		if oldExtensionObj.GetExtensionStatus().GetState() == nil {
+			enqueue(queue, newObj)
+		}
+
+		if bytes.Equal(newExtensionObj.GetExtensionStatus().GetState().Raw, oldExtensionObj.GetExtensionStatus().GetState().Raw) {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	queue.Add(key)
+}
+
+func (e *ExtensionStateSyncController) updateShootState(kind string, namespace string, state *runtime.RawExtension) (reconcile.Result, error) {
+	shootRetriever, err := newShootRetriever()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	shoot, err := shootRetriever.fromCluster(namespace, e.extensionsSeedClient)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	shootState, err := e.k8sGardenClient.GardenCore().CoreV1alpha1().ShootStates(shoot.GetNamespace()).Get(shoot.GetName(), v1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if shootState == nil {
+		e.log.Infof("ShootState resource with name %s doesn't exist in namespace %s", shoot.GetName(), shoot.GetNamespace())
+		return reconcile.Result{RequeueAfter: time.Minute * 10}, nil
+	}
+
+	extensionState := helper.GetExtensionStateResource(shootState.Spec.Extensions, kind)
+	if extensionState == nil {
+		shootState.Spec.Extensions = append(shootState.Spec.Extensions, gardencorev1alpha1.ExtensionResourceState{
+			Kind:  kind,
+			State: gardencorev1alpha1.ProviderConfig{RawExtension: *state},
+		})
+	} else {
+		extensionState.State = gardencorev1alpha1.ProviderConfig{RawExtension: *state}
+	}
+
+	shootState, err = e.k8sGardenClient.GardenCore().CoreV1alpha1().ShootStates(shoot.GetNamespace()).Update(shootState)
+	if err != nil {
+		message := fmt.Sprintf("Shoot %s extension state was NOT succcessfuly synced", kind)
+		e.log.Error(message)
+		e.recorder.Event(shoot, corev1.EventTypeNormal, "ScheduledNextSync", message)
+		return reconcile.Result{RequeueAfter: time.Minute * 1}, err
+	}
+
+	message := fmt.Sprintf("Shoot %s extension state was succcessfuly synced", kind)
+	e.log.Info(message)
+	e.recorder.Event(shoot, corev1.EventTypeNormal, "ScheduledNextSync", message)
+	return reconcile.Result{RequeueAfter: time.Minute * 10}, nil
+}
+
+// ShootRetriever decodes raw Shoots into objects
+type ShootRetriever struct {
+	decoder runtime.Decoder
+}
+
+func newShootRetriever() (*ShootRetriever, error) {
```

Comment:
No need to execute this each and every time. Put this inside a `init()` function, for example.